package/network/services/openvpn/files/lib/netifd/proto/openvpn.sh

```diff
@@ -0,0 +1,167 @@
+
+	# append params
+	append_params "$s" \
+		askpass auth auth_retry auth_user_pass auth_user_pass_verify bcast_buffers ca cert \
+		chroot cipher client_config_dir client_connect client_disconnect comp_lzo connect_freq \
+		connect_retry connect_timeout connect_retry_max crl_verify dev dev_node dev_type dh \
+		echo engine explicit_exit_notify fragment group hand_window hash_size \
+		http_proxy http_proxy_option http_proxy_timeout ifconfig ifconfig_pool \
+		ifconfig_pool_persist ifconfig_push inactive ipchange iroute keepalive \
+		key key_method keysize learn_address link_mtu lladdr local log log_append \
+		lport management management_log_cache max_clients \
+		max_routes_per_client mode mssfix mtu_disc mute nice ns_cert_type ping \
+		ping_exit ping_restart pkcs12 plugin port port_share prng proto rcvbuf \
+		redirect_gateway remap_usr1 remote remote_cert_eku remote_cert_ku remote_cert_tls \
+		reneg_bytes reneg_pkts reneg_sec \
+		replay_persist replay_window resolv_retry route route_delay route_gateway \
+		route_metric rport script_security secret server server_bridge setenv shaper sndbuf \
+		socks_proxy status status_version syslog tcp_queue_limit tls_auth \
+		tls_cipher tls_remote tls_timeout tls_verify tmp_dir topology tran_window \
+		tun_mtu tun_mtu_extra txqueuelen user verb down push up \
+		ifconfig_ipv6 route_ipv6 server_ipv6 ifconfig_ipv6_pool ifconfig_ipv6_push iroute_ipv6 \
+		verify_x509_name
+		
+	return 0
+}
+
+proto_openvpn_init_config() {
+	available=1
+	no_device=1
+	no_proto_task=1
+}
+
+proto_openvpn_setup() {
+	local cfg="$1"
+	local iface="$2"
+
+	local enabled remote
+
+	uci -q get "openvpn.$cfg" || {
+		echo "openvpn[$$]" "openvpn section \"$cfg\" not found"
+		proto_notify_error "$cfg" OPENVPN_CONFIG_NOT_FOUND
+		proto_block_restart "$cfg"
+		return 1
+	}
+
+	config_load 'openvpn'
+	config_get_bool enabled "$1" 'enabled' 0
+	[ "${enabled}" -eq 0 ] || {
+		echo "openvpn[$$]" "Section \"$s\" is enabled in /etc/config/openvpn. It must be disabled to be started with netifd"
+		return 1
+	}
+
+	config_get remote "$cfg" 'remote'
+	remote=${remote%% *}
+	[ -n "$remote" ] && {
+		for ip in $(resolveip -t 5 "$remote"); do
+			#proto_add_host_dependency "$cfg" "$ip"
+			serv_addr=1
+		done
+		[ -n "$serv_addr" ] || {
+			echo "openvpn[$$]" "Could not resolve server address \"$remote\""
+			sleep 5
+			proto_setup_failed "$cfg"
+			return 1
+		}
+	}
+
+	INTERFACE=$cfg
+	prepare_config "$cfg" || {
+		echo "openvpn[$$]" "openvpn config failed"
+		proto_notify_error "$cfg" OPENVPN_CONFIG_FAILED
+		proto_block_restart "$cfg"
+		return 1
+	}
+
+	proto_run_command "$cfg" /usr/sbin/openvpn \
+		--syslog "openvpn($cfg)" \
+		--status "/var/run/openvpn.${cfg}.status" \
+		--cd /etc/openvpn \
+		--config "/var/etc/openvpn-${cfg}.conf" \
+		--setenv INTERFACE $cfg \
+		--route-noexec \
+		--script-security 3 \
+		--route-up /lib/netifd/openvpn-up \
+		--route-pre-down /lib/netifd/openvpn-down
```

Comment:
To keep the style similar to other protocols I'd also like if we'd support an optional `option ifname` and map it to OpenVPN's `--dev` argument.

Along with the proposed config handling above, a working approach could look like:

	# find effective config
	local config
	json_get_vars config
	
	if [ -n "$config" -a ! -f "$config" ]; then
		echo "openvpn[$$]" "User specified configuration '$config' does not exist"
		proto_setup_failed "$cfg"
		return 1
	fi
	
	if [ ! -n "$config" -a -f "/etc/openvpn/$cfg.conf" ]; then
		config="/etc/openvpn/$cfg.conf"
	fi
	
	get_config_param() {
		local variable="$1"
		local option="$2"
		local default="$3"
	
		# if a config file exists, try to find param in it
		if [ -n "$config" ]; then
			local value="$(sed -rne 's/^[ \t]*'"$option"'[ \t]+([^ \t]+)[ \t]*$/\1/p' "$config" | tail -n1)"
			export -n "$variable=${value:-$default}"
	
		# ... else try to look it up in uci
		else
			config_get value "$cfg" "${option//-/_}" "$default"
		fi
	}
	
	local ifname dev_type dev_name
	
	# get ifname from protocol config
	json_get_vars ifname
	
	# fall back to dev from config if no ifname specified
	[ -n "$ifname" ] || get_config_param ifname "$config" dev
	
	# check if device name has been set in passed or generated config
	get_config_param dev_type "$config" dev-type
	
	# guess device type, fall back to "tun"
	case "${dev_type:-${ifname:-$dev_name}}" in
		tap*) dev_type="tap" ;;
		*) dev_type="tun" ;;
	esac
	
	proto_run_command "$cfg" /usr/sbin/openvpn \
		--syslog "openvpn($cfg)" \
		--status "/var/run/openvpn.${cfg}.status" \
		--cd /etc/openvpn \
		--config "${config:-/var/etc/openvpn-${cfg}.conf}" \
		${dev_type:+--dev-type "$dev_type"} \
		${ifname:+--dev "$ifname"} \
		--setenv INTERFACE $cfg \
		--route-noexec \
		--script-security 3 \
		--route-up /lib/netifd/openvpn-up \
		--route-pre-down /lib/netifd/openvpn-down